src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,309 @@
+
+    /**
+     * If a trial is finished, release the connection resource
+     * @param trial remote machine trial job detail
+     */
+    public releaseTrialResource(environment: EnvironmentInformation): void {
+        const executorManager = this.environmentExecutorManagerMap.get(environment.id);
+        if (executorManager === undefined) {
+            throw new Error(`ExecutorManager is not assigned for environment ${environment.id}`);
+        }
+        if (this.machineScheduler === undefined) {
+            throw new Error(`machineScheduler is not initialized!`);
+        }
+        // Note, it still keep reference in trialExecutorManagerMap, as there may be following requests from nni manager.
+        executorManager.releaseExecutor(environment.id);
+        const remoteEnvironment: RemoteMachineEnvironmentInformation = environment as RemoteMachineEnvironmentInformation;
+        if (remoteEnvironment.rmMachineMeta === undefined) {
+            throw new Error(`${remoteEnvironment.id} rmMachineMeta not initialized!`);
+        }
+        this.machineScheduler.recycleMachineReservation(remoteEnvironment.rmMachineMeta);
+    }
+
+    public async startEnvironment(environment: EnvironmentInformation): Promise<void> {
+        if (this.sshConnectionPromises.length > 0) {
+            await Promise.all(this.sshConnectionPromises);
+            this.log.info('ssh connection initialized!');
+            // set sshConnectionPromises to [] to avoid log information duplicated
+            this.sshConnectionPromises = [];
+            if (this.trialConfig ===  undefined) {
+                throw new Error("trial config not initialized!");
+            }
+            this.machineScheduler = new MachineScheduler(Array.from(this.machineExecutorManagerMap.keys()));
+        }
+        const remoteEnvironment: RemoteMachineEnvironmentInformation = environment as RemoteMachineEnvironmentInformation;
+        remoteEnvironment.status = 'WAITING';
+        await this.prepareEnvironment(remoteEnvironment);
+        await this.launchEnvironmentOnScheduledMachine(remoteEnvironment);
+    }
+
+    private async prepareEnvironment(environment: RemoteMachineEnvironmentInformation): Promise<boolean> {
+        const deferred: Deferred<boolean> = new Deferred<boolean>();
+        if (this.trialConfig === undefined) {
+            throw new Error('trial config is not initialized');
+        }
+        if (this.machineScheduler === undefined) {
+            throw new Error('machineScheduler is not initialized');
+        }
+
+        // get an executor from scheduler
+        const rmMachineMeta: RemoteMachineMeta | undefined = this.machineScheduler.scheduleMachine();
+        if (rmMachineMeta === undefined) {
+            this.log.warning(`No available machine!`);
+            deferred.resolve(false);
+        } else {
+            environment.rmMachineMeta = rmMachineMeta;
+            const copyExpCodeDirPromise = this.machineCopyExpCodeDirPromiseMap.get(
```

Comment:
Is it still useful?

Reply:
yes, this promise object is used to copy data to remote machine, and initialized in setClusterMetadata() step. need to use this object to workaround, otherwise setClusterMetadata will cost much time.

Reply:
couldn't find setClusterMetadata, is it replaced by sshConnectionPromises?

Reply:
removed